runelite-client/src/main/java/net/runelite/client/plugins/mta/AlchemyRoom.java

```diff
@@ -0,0 +1,394 @@
+					GameObject cupboard = cupboards.get(i);
+					if (cupboard != null)
+					{
+						client.setHintArrow(cupboard.getWorldLocation());
+						break;
+					}
+				}
+			}
+		}
+	}
+
+
+	@Subscribe
+	public void onGameObjectSpawned(GameObjectSpawned event)
+	{
+		if (!inside())
+			return;
+
+		GameObject spawn = event.getGameObject();
+		for (int i = 0; i < CUPBOARDS.size(); i++)
+		{
+			int id = CUPBOARDS.get(i);
+			if (spawn.getId() == id || spawn.getId() - 1 == id)
+			{
+				cupboards.add(i, spawn);
+			}
+		}
+
+	}
+
+	@Subscribe
+	public void onGameObjectDespawned(GameObjectDespawned event)
+	{
+		if (!inside())
+			return;
+
+		GameObject spawn = event.getGameObject();
+		for (int i = 0; i < CUPBOARDS.size(); i++)
+		{
+			int id = CUPBOARDS.get(i);
+			if (spawn.getId() == id || spawn.getId() - 1 == id)
+			{
+				cupboards.remove(i);
+			}
+		}
+	}
+
+	@Subscribe
+	public void onChatMessage(ChatMessage wrapper)
+	{
+		if (!inside() || !getConfig().alchemy())
+			return;
+
+		String message = wrapper.getMessage();
+		if (wrapper.getType() == ChatMessageType.SERVER)
+		{
+			if (message.contains(YOU_FOUND))
+			{
+				String item = message.replace(YOU_FOUND, "").trim();
+				fill(getClicked(), item);
+			}
+			else if (message.equals(EMPTY))
+			{
+				locations[getClicked()] = AlchemyItem.EMPTY;
+			}
+		}
+	}
+
+	private void reset()
+	{
+		reset(false);
+	}
+
+	private void reset(boolean logout)
+	{
+		for (int i = 0; i < locations.length; i++)
+		{
+			locations[i] = AlchemyItem.UNKNOWN;
+		}
+
+		if (logout)
+		{
+			for (int i = 0; i < cupboards.size(); i++)
+			{
+				cupboards.set(i, null);
+			}
+		}
+	}
+
+	@Override
+	public boolean inside()
+	{
+		return client.getWidget(WidgetID.MTA_ALCHEMY_GROUP_ID, 0) != null;
+	}
+
+	public String getBest()
+	{
+		for (int i = 0; i < INFO_START; i++)
+		{
+			int index = i + INFO_START;
+
+			Widget textWidget = client.getWidget(WidgetID.MTA_ALCHEMY_GROUP_ID, index);
+			if (textWidget == null)
+			{
+				reset();
+				return "";
+			}
+
+			String item = textWidget.getText().replace(":", "");
+
+			Widget pointsWidget = client.getWidget(WidgetID.MTA_ALCHEMY_GROUP_ID, index + INFO_START);
+			int points = Integer.parseInt(pointsWidget.getText());
+			if (points == BEST_POINTS)
+			{
+				return item;
+			}
+		}
+
+		return "";
+	}
+
+	private int getClicked()
+	{
+		GameObject nearest = null;
+		double distance = Double.MAX_VALUE;
+
+		for (GameObject object : cupboards)
+		{
+			if (object == null)
+			{
+				continue;
+			}
+
+			WorldPoint mine = client.getLocalPlayer().getWorldLocation();
+			double objectDistance = object.getWorldLocation().distanceTo(mine);
+			if (nearest == null || distance > objectDistance)
+			{
+				nearest = object;
+				distance = objectDistance;
+			}
+		}
+
+		return cupboards.indexOf(nearest);
+	}
+
+	private void fill(int index, String item)
+	{
+		int start = AlchemyItem.indexOf(item);
+		int distance = 8 - index;
+		start = (start + distance) % 8;
```

Comment:
Maybe declare distance first, and then int start = (AlchemyItem.indexOf(item)+ distance) % 8;